refactor: make checkqueue manage the threads by itself (also removed some boost dependencies)

ken2812221 opened the pull request:
This PR removes use of the following object in checkqueue
- `boost::condition_variable`
- `boost::mutex`
- `boost::unique_lock`

Instead of using `boost::thread_group::interrupt_all` to trigger `boost::thread_interrupted` exception, it should call `InterruptScriptCheck()` or `scriptcheckqueue.Interrupt()` to interrupt the checkqueue thread.

Also, it adds three functions:
- `StartScriptCheck()`
- `InterruptScriptCheck()`
- `StopScriptCheck()`

for clearer logic for script check threads.

practicalswift commented on test/lint/lint-includes.sh:
```diff
@@ -68,7 +68,6 @@ EXPECTED_BOOST_INCLUDES=(
     boost/test/unit_test.hpp
     boost/thread.hpp
     boost/thread/condition_variable.hpp
```
According to PR title `boost/thread/condition_variable.hpp` should go as well? :-)

ken2812221 replied:
I only remove them in checkqueue, the another place to use it is in scheduler.

practicalswift commented:
Concept ACK

Towards a Boost free future! :-)

Empact commented on src/checkqueue.h:
```diff
@@ -62,17 +62,19 @@ class CCheckQueue
     //! The maximum number of elements to be processed in one batch
     unsigned int nBatchSize;
 
+    std::atomic<bool> fInterrupted;
```
nit: could just call this `interrupted`

Empact commented on src/checkqueue.h:
```diff
@@ -96,7 +98,11 @@ class CCheckQueue
+                        return false;
                     cond.wait(lock); // wait
+                    if (interrupted)
+                        return false;
```
> If an `if` only has a single-statement `then`-clause, it can appear
    on the same line as the `if`, without braces. In every other case,
    braces are required, and the `then` and `else` clauses must appear
    correctly indented on a new line.
https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c

ch4ot1c commented on src/validation.h:
```diff
@@ -260,8 +260,12 @@ bool LoadBlockIndex(const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs
-void ThreadScriptCheck();
+/** Start script checking thread */
+void StartScriptCheck();
+/** Interrupt all script checking thread */
```
Pluralize to `threads`

ch4ot1c commented:
utACK

sipa commented:
utACK 9aacef638f0c2159930640762daf524942052a9d

donaloconnor commented:
utACK https://github.com/bitcoin/bitcoin/pull/14464/commits/9aacef638f0c2159930640762daf524942052a9d

donaloconnor commented on src/checkqueue.h:
```diff
@@ -157,6 +161,16 @@ class CCheckQueue
+            std::lock_guard<std::mutex> lock(mutex);
+            interrupted = true;
+        }
+        condMaster.notify_all();
```
tiny nit: only one master thread so maybe `condMaster.notify_one()` is more consistent. No difference to code though.

ken2812221 commented:
```diff
git diff 9aacef638f0c2159930640762daf524942052a9d 68e50c8779ae94456ceace4d488b4f07e06e4ad7
diff --git a/src/checkqueue.h b/src/checkqueue.h
index 6d158ce99..015811f4d 100644
--- a/src/checkqueue.h
+++ b/src/checkqueue.h
@@ -167,7 +167,7 @@ public:
             std::lock_guard<std::mutex> lock(mutex);
             interrupted = true;
         }
-        condMaster.notify_all();
+        condMaster.notify_one();
         condWorker.notify_all();
     }

diff --git a/src/test/test_bitcoin.cpp b/src/test/test_bitcoin.cpp
index a3f01a6b3..fb51bf011 100644
--- a/src/test/test_bitcoin.cpp
+++ b/src/test/test_bitcoin.cpp
@@ -118,8 +118,8 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
 TestingSetup::~TestingSetup()
 {
     InterruptScriptCheck();
-    StopScriptCheck();
     threadGroup.interrupt_all();
+    StopScriptCheck();
     threadGroup.join_all();
     GetMainSignals().FlushBackgroundCallbacks();
     GetMainSignals().UnregisterBackgroundSignalScheduler();

practicalswift commented on src/test/checkqueue_tests.cpp:
```diff
@@ -149,9 +148,9 @@ typedef CCheckQueue<FrozenCleanupCheck> FrozenCleanup_Queue;
 {
     auto small_queue = MakeUnique<Correct_Queue>(QUEUE_BATCH_SIZE);
-    boost::thread_group tg;
+    std::vector<std::thread> tg;
```
nit: Request vector capacity of `nScriptCheckThreads` before loop?

practicalswift commented on src/test/transaction_tests.cpp:
```diff
@@ -462,12 +462,13 @@ BOOST_AUTO_TEST_CASE(test_big_witness_transaction) {
     // check all inputs concurrently, with the cache
     PrecomputedTransactionData txdata(tx);
-    boost::thread_group threadGroup;
+    std::vector<std::thread> threadGroup;
```
nit: Same here. Request a capacity of `20`? :-)

JeremyRubin commented on src/checkqueue.h:
```diff
@@ -96,7 +98,9 @@ class CCheckQueue
                     }
                     nIdle++;
+                    if (interrupted) return false;
                     cond.wait(lock); // wait
```
Add the condition as a callback to wait.

e.g., cond.wait(lock, [&] { return interrupted || !queue.empty(); })

JeremyRubin reviewed (commented):
Weak concept Ack.

Necessary step to get rid of boost, but conflicts with some unmerged work (https://github.com/bitcoin/bitcoin/pull/9938, https://github.com/JeremyRubin/bitcoin/pull/1) which does the same thing.

Overall, this appears to be correct.

Before merge I'd like to see some benchmarks, the checkqueue code is very performance sensitive, sometimes the boost condition variables and mutexs difference from std's can be substantial. (I'm not kidding -- the time to wake-up and condition variable sleep behavior is one of the major performance issues with this code).

sipa commented on src/init.cpp:
```diff
@@ -1199,12 +1201,6 @@ bool AppInitMain()
     InitSignatureCache();
     InitScriptExecutionCache();
 
-    LogPrintf("Using %u threads for script verification\n", nScriptCheckThreads);
```
I was expecting to see this replaced with a `StartScriptCheck();` line; why is that not necessary?

ken2812221 replied:
I think I forgot to add it. Fixed now.

ken2812221 commented:
@JeremyRubin
benchmark result on my virtual machine:
master 816fab9ccae568612d5ed90378b4587256925a1e
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 83.1845, 0.000461328, 0.000692877, 0.000668382
```
This PR 625c55a70c376af9a94a2bb118c39685d3b82b33
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 66.268, 0.000451829, 0.000491473, 0.000473479
```

JeremyRubin commented:
That's great, thanks for checking that -- much bigger change than I expected it to be.

Looking at the numbers, it seems not only to be faster on the microbenchmark, but more consistent.

This merits running two additional tests, if you have energy for it:

1. Time to reindex (with assumevalid up to last month or something so it doesn't take forever
2. Two nodes (one with this PR one without) connected behind a third node (on latest release), with debug bench set. Collect the logs on ConnectBlock.
     - Also collect a general system profile of which processes are using what on the cpu. 

The second test is important because one thing I noticed when working on this code previously is that it was trivially easy to get speedups on this code at the expense of general performance. E.g., getting rid of condition variables (and just spinning until jobs came in) was much faster, but meant that the OS scheduler would be giving our CheckQueue workers run-time when there was nothing to do, impacting other processes on the computer (most likely because of causing more cache invalidation, rather than the impact of losing a few cycles of a process switch).  Hence my suspicion: the 20% speedup is because the standard condition variables are not using system sleep (which plays better with the OS). This can actually negatively impact the process too as our main thread might be getting context switched out while we are trying to add jobs or something rather than waiting for notify.

JeremyRubin commented on src/checkqueue.h:
```diff
@@ -90,13 +92,13 @@ class CCheckQueue
                     }
                     nIdle++;
-                    cond.wait(lock); // wait
+                    cond.wait(lock, [&] { return interrupted || !queue.empty() || fMaster && nTodo == 0; }); // wait
```
I think you could assert(! (fMaster && nTodo == 0)) here.

If fMaster were *ever* true here, we would deadlock.

Hence, there is no point to check that...

JeremyRubin commented:
This version is not correct either if I'm correct about the block getting flushed with validity cached during shutdown.

Now you might incorrectly cache that a block was valid when it may not have been.

You need to finish processing the block, then shut down. Or, if a shutdown is triggered, abandon the block in the master (you could make a  trivalent return: false, true, shutdown).

JeremyRubin commented:
Actually I think that's not right, I think this version will just hang because nTodo will be != 0 and queue.empty() will be true, so the master would sleep and never get woken?

ken2812221 commented:
@JeremyRubin master would never wait/sleep if `interrupted` is true.

ken2812221 commented:
@JeremyRubin
How about now? I think the logic here would be more clear. It would keep checking until finish.
If there are new tasks added into queue after the worker threads ended, it would be only master worker checking the tasks unfortunately. However I think this is better than interrupt the master thread.

JeremyRubin commented:
I still think it's best to make control of the checkqueue (as opposed to participation) protected by a mutex.

Easiest to audit/review.

But do let me know if that fixes the corruption issue...

ken2812221 commented:
> I still think it's best to make control of the checkqueue (as opposed to participation) protected by a mutex.

There already have `ControlMutex` to ensure that there is only one `CCheckQueueControl` to access the CCheckQueue.

JeremyRubin commented:
I am saying that a simple change to make is to have Interrupt take the ControlMutex.

ken2812221 commented:
>I am saying that a simple change to make is to have Interrupt take the ControlMutex.

That would cause recursive lock in test_bitcoin. Also, I don't think that it's necessary to lock ControlMutex.

JeremyRubin commented:
Where's the recursive lock?

ken2812221 commented:
In test_bitcoin, it constuct `CCheckQueueControl` and call `Interrupt` in same thread

JeremyRubin commented:
Interrupt should be called in the destructor of the test which should happen after any queuecontrols are destructed though?

ken2812221 commented:
> Interrupt should be called in the destructor of the test which should happen after any queuecontrols are destructed though?

I think it would be better to make the checkqueue to control their worker threads instead of calling `Thread()` in somewhere.

ken2812221 committed:
refactor: Make CCheckQueue interruptible

This commit add a Interrupt function for CCheckQueue that it can handle interrupt by itself instead of relying on boost thread interrupt

ken2812221 committed:
refactor: Drop boost in CCheckQueue

replace boost::mutex with debuggable Mutex
replace boost::condition_variable with std::condition_variable
add const specifier to fMaster and nBatchSize
add clang thread safety attributes
move init value of member of CCheckQueue from constructor to definition

ken2812221 committed:
refactor: Make CCheckQueue worker threads be owned by itself

Manage CCheckQueue workder threads by itself. Do not expose the threads outside the object.

ken2812221 commented:
Rebased

ken2812221 commented:
Rebased

minblock commented:
Can I freebase it?

ken2812221 commented:
rebased

MarcoFalke commented:
Could try a `git rebase -Xignore-all-space df36ddf9ce8a4dcf0d7f324e57a13abb6cf6a57c`?

practicalswift commented:
@ken2812221 Why the close? :-)

ken2812221 commented:
The CI was failed and I don't have time to find the bugs out recently, so I closed it, I might reopen this when I has more time working on this.
